graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,255 @@
+            if (throwable instanceof GraphQLError) {
+                graphQLErrors.add((GraphQLError) throwable);
+            } else {
+                graphQLErrors.add(GraphqlErrorBuilder.newError()
+                        .message(throwable.getMessage())
+                        .build());
+            }
+        }
+
+        if (result instanceof DataFetcherResult<?>) {
+            DataFetcherResult<?> theResult = (DataFetcherResult) result;
+            if (theResult.hasErrors()) {
+                graphQLErrors.addAll(theResult.getErrors());
+            }
+        }
+
+        return graphQLErrors;
+    }
+
+    /**
+     * Stores timing information and a map of nodes for field information.
+     */
+    private static class FederatedTracingState implements InstrumentationState {
+        private final Instant startRequestTime;
+        private final long startRequestNanos;
+        private final LinkedHashMap<ExecutionPath, Reports.Trace.Node.Builder> nodesByPath;
+
+        private FederatedTracingState() {
+            // record start time when creating instrumentation state for a request
+            startRequestTime = Instant.now();
+            startRequestNanos = System.nanoTime();
+
+            nodesByPath = new LinkedHashMap<>();
+            nodesByPath.put(ExecutionPath.rootPath(), Reports.Trace.Node.newBuilder());
+        }
+
+        @NotNull
+        Reports.Trace toProto() {
+            return Reports.Trace.newBuilder()
+                    .setStartTime(getStartTimestamp())
+                    .setEndTime(getNowTimestamp())
+                    .setDurationNs(getDuration())
+                    .setRoot(nodesByPath.get(ExecutionPath.rootPath()))
+                    .build();
+        }
+
+        /**
+         * Adds node to nodesByPath and recursively ensures that all parent nodes exist.
+         */
+        void addNode(ExecutionStepInfo stepInfo, long startFieldNanos, long endFieldNanos, List<GraphQLError> errors, SourceLocation fieldLocation) {
+            ExecutionPath path = stepInfo.getPath();
+            Reports.Trace.Node.Builder parent = getParentNode(path);
+
+            Reports.Trace.Node.Builder node = parent.addChildBuilder()
+                    .setStartTime(startFieldNanos)
+                    .setEndTime(endFieldNanos)
+                    .setParentType(simplePrint(stepInfo.getParent().getUnwrappedNonNullType()))
+                    .setType(stepInfo.simplePrint())
+                    .setResponseName(stepInfo.getResultKey());
+
+            String originalFieldName = stepInfo.getField().getName();
+
+            // set originalFieldName only when a field alias was used
+            if (!originalFieldName.equals(stepInfo.getResultKey())) {
+                node.setOriginalFieldName(originalFieldName);
+            }
+
+            errors.forEach(error -> {
+                Reports.Trace.Error.Builder builder = node.addErrorBuilder().setMessage(error.getMessage());
```

Comment:
The JS implementation also takes other kinds of errors and puts them on the root node — most notably parse and validation errors, which the instrumentation can instrument. Do you want to implement that too?

Reply:
Great idea, this was pretty easy. I'm not entirely sure if I'm grabbing locations off the throwable in the best way (in `convertErrors`) ... got any other ideas?

Reply:
Looks good enough to me